### ISSUE TITLE
Do not detect US_USELESS_SUPPRESSION_ON_* warnings on Generated code.

### ISSUE THREAD

**Issue by SimSonic** [spotbugs/spotbugs#3350]
In my project I have lombok which is set to add `@SuppressFBWarnings` onto all generated constructors/methods thought it's configuration:

```
// root lombok.config
lombok.extern.findbugs.addSuppressFBWarnings = true
```

After updating to SpotBugs 4.9.3 javac throws medium warnings `US_USELESS_SUPPRESSION_ON_*` on generated code.

I believe we should have some possibility to prevent generated code from being scanned.

May be disable all warnings including these ones on all code which annotated as `@Generated` (javax/jakarta?).

**Comment by fenuks** [comment]
SpotBugs have undocumented (it is not mentioned in
https://spotbugs.readthedocs.io/en/latest/filter.html) Annotation filter.
Unfortunately, as far I can see, it only works with annotations on classes.
Lombok adds `@lombok.Generated` on generated methods, so I think extending
[AnnotationMatcher](https://github.com/spotbugs/spotbugs/blob/master/spotbugs/src/main/java/edu/umd/cs/findbugs/filter/AnnotationMatcher.java) to support annotations on methods would be one way to solve this problem.

**Comment by gtoison** [comment]
Thank you for reporting the issue
@SimSonic could you please share a few example where the issue happens? This will help making sure that the problem is understood correctly

**Comment by SimSonic** [comment]
@gtoison yes, here is a minimal example with maven, Java 21, based on Spring Initialized application:

[issue-3350-demo.zip](https://github.com/user-attachments/files/19244992/issue-3350-demo.zip)

Just run `mvn clean verify` to see this:

```
[ERROR] Medium: Suppressing annotation on the method new com.example.demo.DemoApplication$BrokenClass() is unnecessary [com.example.demo.DemoApplication$BrokenClass] At DemoApplication.java:[line 10] US_USELESS_SUPPRESSION_ON_METHOD
[ERROR] Medium: Suppressing annotation on the method com.example.demo.DemoApplication$BrokenClass.hashCode() is unnecessary [com.example.demo.DemoApplication$BrokenClass] At DemoApplication.java:[line 10] US_USELESS_SUPPRESSION_ON_METHOD
[ERROR] Medium: Suppressing annotation on the method com.example.demo.DemoApplication$BrokenClass.toString() is unnecessary [com.example.demo.DemoApplication$BrokenClass] At DemoApplication.java:[line 10] US_USELESS_SUPPRESSION_ON_METHOD
```

**Comment by rougou** [comment]
Is there a way to turn off this feature all together? I'm using spotbugs-maven-plugin and adding `<omitVisitors>NoteSuppressedWarnings</omitVisitors>` to the configuration doesn't seem to have any effect.
For one of my projects, spotbugs execution time went from 1 minute to a whopping 24 minutes starting from version 4.9.2.
After deleting `lombok.extern.findbugs.addSuppressFBWarnings=true` it went down to 2 minutes and 39 seconds, which is better but still not ideal.

**Comment by JuditKnoll** [comment]
> For one of my projects, spotbugs execution time went from 1 minute to a whopping 24 minutes starting from version 4.9.2. After deleting `lombok.extern.findbugs.addSuppressFBWarnings=true` it went down to 2 minutes and 39 seconds, which is better but still not ideal.

@rougou Thanks for highlighting this issue. This execution time change was between version 4.9.1 and 4.9.2, right? Can you please try run SpotBugs version 4.9.3 on the mentioned project? Do you get similar execution time to 4.9.2?

@gtoison Can you please take a look at this?

> Is there a way to turn off this feature all together? I'm using spotbugs-maven-plugin and adding `<omitVisitors>NoteSuppressedWarnings</omitVisitors>` to the configuration doesn't seem to have any effect. 

In 4.9.3 `UselessSuppressionDetector` got introduced for this functionality instead of `NoteSuppressedWarnings`. On that version you should be able to omit that visitor.

**Comment by gtoison** [comment]
`UselessSuppressionDetector ` is a fake detector, so even if it omitted the logic to look for check for unnecessary suppression would still run. I don't anticipate that it would make a difference, but I could be wrong, it's worth a try.

@rougou it would be great if you could share a few thread dumps of the JVM running SpotBugs (or better yet attach a profiler to it), so we can see where all this time is spent.

**Comment by rougou** [comment]
@JuditKnoll @gtoison
Yes, I tried omitting both NoteSuppressedWarnings and UselessSuppressionDetector, but neither seemed to have any effect.
However, specifying a different detector with the `visitors` option does seem to disable the useless suppression checker (along with every other detector not specified).

Also, on second glance, spotbugs itself finishes pretty quickly. It seems the spotbugs maven plugin is just having a hard time generating the xml with so many issues (the file comes out to around 50mb). So I can work around this problem by excluding the specific patterns from the report.

Sorry for the confusion.

**Comment by JuditKnoll** [comment]
#3367 got merged to master, which fixes this issue. It will be released in SpotBugs 4.9.4.

**Comment by programista-zacny** [comment]
> [#3367](https://github.com/spotbugs/spotbugs/pull/3367) got merged to master, which fixes this issue. It will be released in SpotBugs 4.9.4.

@JuditKnoll When do you plan to release 4.9.4? I think a few people are waiting for this fix :)

**Comment by JuditKnoll** [comment]
> > [#3367](https://github.com/spotbugs/spotbugs/pull/3367) got merged to master, which fixes this issue. It will be released in SpotBugs 4.9.4.
> 
> [@JuditKnoll](https://github.com/JuditKnoll) When do you plan to release 4.9.4? I think a few people are waiting for this fix :)

We run into an issue blocking the release. You can find the latest status info in [this discussion](https://github.com/spotbugs/spotbugs/discussions/3380). Thank you for your patience until then.

**Comment by SimSonic** [comment]
Latest MapStruct 1.6.3 is using `import javax.annotation.processing.Generated;` ...

**Comment by gtoison** [comment]
SpotBugs 4.9.4 is now released!
